File: C++/7 - Binary Search - Greg Hogg DSA Course Materials Lecture 7.cpp
// Binary Search - Greg Hogg DSA Course Materials Lecture 7

#include <iostream>
#include <vector>

// Traditional Binary Search - O(log n) time, O(1) space
bool binarySearch(const std::vector<int>& arr, int target) {
    int L = 0;
    int R = arr.size() - 1;

    while (L <= R) {
        int M = L + (R - L) / 2;
        if (arr[M] == target) {
            return true;
        } else if (target < arr[M]) {
            R = M - 1;
        } else {
            L = M + 1;
        }
    }
    return false;
}

// Condition-based Binary Search - O(log n) time, O(1) space
int binarySearchCondition(const std::vector<bool>& arr) {
    int L = 0;
    int R = arr.size() - 1;

    while (L < R) {
        int M = L + (R - L) / 2;
        if (arr[M]) {
            R = M;
        } else {
            L = M + 1;
        }
    }
    return L;
}

int main() {
    std::vector<int> A = {-3, -1, 0, 1, 4, 7};
    int target = 8;
    bool result = binarySearch(A, target);
    std::cout << "Is " << target << " in the array? " << (result ? "Yes" : "No") << std::endl;

    std::vector<bool> B = {false, false, false, false, false, true, true};
    int conditionResult = binarySearchCondition(B);
    std::cout << "First true element index: " << conditionResult << std::endl;

    return 0;
}

File: Python/7 - Binary Search - Greg Hogg DSA Course Materials Lecture 7.py
# Binary Search - Greg Hogg DSA Course Materials Lecture 7

# The Colab Notebook can be viewed here: https://colab.research.google.com/drive/1eN3A8i88C641wlIdwTeahwoQLcthCGtD

A = [-3, -1, 0, 1, 4, 7]

# Naive O(n) searching
if 8 in A:
  print(True)

# Traditional Binary Search - Looking up if number is in array:
# Time: O(log n)
# Space: O(1)

def binary_search(arr, target):
  N = len(arr)
  L = 0
  R = N - 1

  while L <= R:
    M = L + ((R-L) // 2)

    if arr[M] == target:
      return True
    elif target < arr[M]:
      R = M - 1
    else:
      L = M + 1

  return False

binary_search(A, 8)

# Condition based
B = [False, False, False, False, False, False, True]

def binary_search_condition(arr):
  N = len(arr)
  L = 0
  R = N - 1

  while L < R:
    M = (L + R) // 2

    if B[M]:
      R = M
    else:
      L = M + 1

  return L

binary_search_condition(B)

